aiida/engine/processes/workchains/workchain.py

```diff
@@ -123,30 +122,32 @@ def on_run(self):
         self.node.set_stepper_state_info(str(self._stepper))
 
     def insert_awaitable(self, awaitable):
-        """
-        Insert a awaitable that will cause the workchain to wait until the wait
-        on is finished before continuing to the next step.
+        """Insert an awaitable that should be terminated before before continuing to the next step.
 
-        :param awaitable: The thing to await
+        :param awaitable: the thing to await
         :type awaitable: :class:`aiida.engine.processes.workchains.awaitable.Awaitable`
         """
         self._awaitables.append(awaitable)
+        status = 'Waiting for child processes: {}'.format(', '.join([str(_.pk) for _ in self._awaitables]))
+        self.node.set_process_status(status)
 
     def remove_awaitable(self, awaitable):
-        """
-        Remove a awaitable.
+        """Remove an awaitable.
 
-        Precondition: must be a awaitable that was previously inserted
+        Precondition: must be an awaitable that was previously inserted.
 
-        :param awaitable: The awaitable to remove
+        :param awaitable: the awaitable to remove
         """
         self._awaitables.remove(awaitable)
 
+        if not self._awaitables:
+            self.node.set_process_status(None)
```

Comment:
Do I understand correctly that:
- every time a new await able is added, the status is added with the list of PKs of th awaitables currently interted
- when one is removed, the status is not changed (unless all have gone)
- Also, if an awaitable is added after a few have been already removed, the string is updated with the currently inserted ones.
Is this the intended behaviour? It seems a bit counterintuitive.

Also, is there the possibility that the last two finish approximatively at the same time and both have a non-empty `self._awaitables` or this is not possible and the whole logic is "thread-safe"?`

Reply:
yes the behavior you describe is correct. The only problematic part there, I think, is where the status is not updated until **all** awaitables are completed. I can change this to update the status each time an awaitable is removed. Your third bullet point scenario is currently not really possible since sub processes won't be found to be finished by the one who launched them, until he tries to go to the next step. But while waiting to go to the next step, he cannot insert. So there is that guarantee that once he added all awaitables and started waiting, no new ones will be added.